rclpy/test/test_clock.py

```diff
@@ -296,6 +330,31 @@ def run():
     assert retval is False
 
 
+@pytest.mark.parametrize('ros_time_enabled', (True, False))
+def test_sleep_for_ros_time_toggled(default_context, ros_time_enabled):
+    clock = ROSClock()
+    clock._set_ros_time_is_active(not ros_time_enabled)
+
+    retval = None
+
+    def run():
+        nonlocal retval
+        retval = clock.sleep_for(Duration(seconds=999999))
+
+    t = threading.Thread(target=run, daemon=True)
+    t.start()
+
+    # wait for thread to get inside sleep_for call
+    time.sleep(0.2)
```

Comment:
I don't like much a test depending on a race condition.
I would rather not have a test if this is the only way.

By the way, I don't get what's the intention of the test.

Reply:
Yeah, it's not my favorite either. Another way may be to inspect the current frame of the other thread to see what code it's exiting, but I don't think I can implement that robustly in a reasonable time.

https://github.com/ros2/rclpy/pull/858#discussion_r768967247

---

The `ros_time_toggled` tests enable or disable ROS time while the method is sleeping. Since that switches from system time to `/clock` time, it's not clear what amount of time should be slept when that happens. This test makes sure the method wakes up and returns False.

Reply:
A reasonable alternative is to toggle the clock type in a loop, wait until `retval is not None` and then assert that `retval is False`.
The loop could have a certain maximum duration, and we could make it "less busy" be sleeping for a bit between each time we toggle the clock type.


e.g.:


```python3
    retval = None

    def run():
        nonlocal retval
        retval = clock.sleep_for(Duration(seconds=999999))

    t = threading.Thread(target=run)
    t.start()

    start = time.now()
    # wait for thread to get inside sleep_for call
    while retval is None and time.now() < start + 0.5:
            clock._set_ros_time_is_active(ros_time_enabled)
            ros_time_enabled = not ros_time_enabled

    assert retval is False
    t.join()
```

Not really amazing, and it doens't make much sense to have the test parameterized with `ros_time_enabled` (as we really don't now which of the toggles made the clock stop waiting), but probably a bit more robust.

Reply:
anyway, I'm okay with the current approach, but we should track CI the first few days to see if the tests are reliable.